### PR TITLE
emacspeak: init at 51.0

### LINKED ISSUE #70261: Emacspeak
**Project description**

Emacspeak is a speech interface that allows visually impaired users to interact independently and efficiently with the computer. 

Audio formatting—a technique pioneered by AsTeR—and full support for W3C's Aural CSS (ACSS) allows Emacspeak to produce rich aural presentations of electronic information. By seamlessly blending all aspects of the Internet such as Web-surfing and messaging, Emacspeak speech-enables local and remote information via a consistent and well-integrated user interface. Available free of cost on the Internet, Emacspeak has dramatically changed how the author and thousands of blind and visually impaired users around the world interact with the personal computer and the Internet. A rich suite of task-oriented tools provides efficient speech-enabled access to the audio desktop and evolving semantic WWW. When combined with Linux running on low-cost PC hardware, Emacspeak/Linux provides a reliable, stable speech-friendly solution that opens up the Internet to visually impaired users around the world.

**Metadata**

* homepage URL: http://emacspeak.sourceforge.net/
* source URL: https://github.com/tvraman/emacspeak
* license: GPL
* platforms: Linux

### PR DESCRIPTION
**Pull request by Dema**
Closes #70261

###### Motivation for this change

New package

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Needs to be tested by users of emacspeak. There're a lot of irrelevant files in the sources like blog posts, speeches, radio playlists and so on. I copied what I think is needed to run it. But I don't know how to use emacspeak, so we need some real user to test if it works. 

###### Notify maintainers

I didn't set `meta.maintainers` field

### CODE REVIEW

**Comment by adisbladis** [comment]
Emacs packages should go in the emacs package set, this one should go in https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/emacs-modes/manual-packages.nix.

File: pkgs/applications/audio/emacspeak/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{ stdenv, fetchgit, rsync, gnused, emacs, tcl, espeak-classic }:
+
+stdenv.mkDerivation rec {
+  pname = "emacspeak";
+  version = "50.0";
+
+  src = fetchgit {
```
Comment (jonringer) [inline pkgs/applications/audio/emacspeak/default.nix]: please use fetchFromGitHub, it's less expensive than pulling down an entire git history

File: pkgs/applications/audio/emacspeak/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  make espeak
+'';
+
+# installPhase = ''
+#   mkdir -p $out/share/emacs/site-lisp/emacspeak/servers/native-espeak
+# '';
+postInstall = ''
+  cd servers/native-espeak
+  make DESTDIR=$out install
+  cd ../../
```
Comment (jonringer) [inline pkgs/applications/audio/emacspeak/default.nix]: This seems odd. Do you have to call a specific Makefile?
```
makeFlags = [ "DESTDIR=${placeholder "out"}" ];
```

Comment (Dema) [reply pkgs/applications/audio/emacspeak/default.nix]: Yeah, there's a  separate Makefile that builds `tclspeak.so`. Main Makefile doesn't do install. It just prints message to add current folder to emacs config. Could you explain a bit about `placeholder`? `$out` is deprecated?

**Comment by Dema** [comment]
Moved `callPackage` to applications/editors/emacs-modes/manual-packages.nix but left default.nix in applications/audio, is it ok?

**Comment by adisbladis** [comment]
Please also move the expression itself.

File: pkgs/applications/audio/emacspeak/default.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/applications/audio/emacspeak/default.nix]: in most situations, there's no difference. From what I understand, placeholder just ensures that a variable doesn't get resolved until the last possible second. I don't remember the exact scenarios in which it was necessary to add the placeholder feature tbh

Comment (jonringer) [reply pkgs/applications/audio/emacspeak/default.nix]: but these three lines could be condensed to:
```
make -C servers/native-espeak DESTDIR=$out install
```

**Comment by Dema** [comment]
done!

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    substituteInPlace servers/native-espeak/Makefile --replace 'LIBPARENTDIR = ''${PREFIX}/'  'LIBPARENTDIR = /'
+'';
+
+  preConfigure = ''
+    make config
+  '';
+
+  postBuild = ''
+  make espeak
+'';
```
Comment (jonringer) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix]: fix alignment please

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  version = "50.0";
+
+  src = fetchFromGitHub {
+    owner = "tvraman";
+    repo = "emacspeak";
+    rev = version;
+    sha256 = "043mn3rr48gx2mi4rnkcvp09fwhs88yyqcszavkzvvd57mh4vd0h";
+  };
+
+  nativeBuildInputs = [rsync gnused] ;
```
Comment (jonringer) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix]:
```suggestion
  nativeBuildInputs = [ rsync gnused ] ;
```

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  rsync -a --exclude-from=ex lisp $out/share/emacs/site-lisp/emacspeak/
+  rsync -a --exclude-from=ex scapes $out/share/emacs/site-lisp/emacspeak/
+  rsync -a --exclude-from=ex utils $out/share/emacs/site-lisp/emacspeak/
+'';
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/tvraman/emacspeak/;
+    description = "Emacs extension that provides spoken output";
+    license = licenses.gpl2;
+    maintainers = [];
```
Comment (jonringer) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix]: do you mind adding yourself as a maintainer? mostly just so you get notified if this gets bumped. Then you can verify that the software still works.

**Comment by Dema** [comment]
Could you, please re-review PR. I got back to nixos after rage quitting it last autumn after fighting with nix language. But it's too attractive in it's core so I decided to return again :-)

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    owner = "tvraman";
+    repo = "emacspeak";
+    rev = version;
+    sha256 = "06azq2566pj4pnyawqm9387ybp3wr1pr6828hhdxhki4dpn9pj8g";
+  };
+
+  nativeBuildInputs = [ rsync gnused ];
+  buildInputs = [ emacs tcl tclx espeak-ng ];
+  postPatch = ''
+    substituteInPlace servers/native-espeak/Makefile --replace '/usr/include/tcl$(TCL_VERSION)' ${tcl}/include
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:17]: You can overwrite both setting:

```
makeFlags = [ "TCL_INCLUDE=${tcl}/include" "LIBPARENTDIR = /" ];
```

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    substituteInPlace servers/native-espeak/Makefile --replace 'LIBPARENTDIR = ''${PREFIX}/'  'LIBPARENTDIR = /'
+  '';
+
+  preConfigure = ''
+    make config
+  '';
+
+  postBuild = ''
+    make espeak
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:26]: makeFlags won't be effective here, but `make espeak` just calls `servers/native-espeak/Makefile`. So you can also do this:

```
make -C servers/native-espeak "TCL_INCLUDE=${tcl}/include" "LIBPARENTDIR = /"
```

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  version = "51.0";
+
+  src = fetchFromGitHub {
+    owner = "tvraman";
+    repo = "emacspeak";
+    rev = version;
+    sha256 = "06azq2566pj4pnyawqm9387ybp3wr1pr6828hhdxhki4dpn9pj8g";
+  };
+
+  nativeBuildInputs = [ rsync gnused ];
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:14]: sed is already in stdenv.

```suggestion
  nativeBuildInputs = [ rsync ];
```

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  buildInputs = [ emacs tcl tclx espeak-ng ];
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:15]: If emacs is used during the build it should be in `nativeBuildInputs`.

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 10]
```diff
@@ -0,0 +1,46 @@
+{ stdenv, fetchFromGitHub, rsync, gnused, emacs, tcl, tclx, espeak-ng }:
+
+stdenv.mkDerivation rec {
+  pname = "emacspeak";
+  version = "51.0";
+
+  src = fetchFromGitHub {
+    owner = "tvraman";
+    repo = "emacspeak";
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:9]: In Archlinux they seem to use the release tarball.
This should contain less irrelevant files: 

https://github.com/tvraman/emacspeak/blob/master/Makefile#L74
https://github.com/tvraman/emacspeak/releases/download/51.0/emacspeak-51.0.tar.bz2
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=emacspeak

**Comment by Dema** [comment]
Updated as per suggestions and took installation code from arch package :), looks a bit cleaner.

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  pname = "emacspeak";
+  version = "51.0";
+
+
+  src = fetchurl {
+    url = "https://github.com/tvraman/emacspeak/releases/download/${version}/${pname}-${version}.tar.bz2";
+    sha256 = "09a0ywxlqa8jmc0wmvhaf7bdydnkyhy9nqfsdqcpbsgdzj6qpg90";
+  };
+
+  nativeBuildInputs = [ makeWrapper rsync emacs ];
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:13]:
```suggestion
  nativeBuildInputs = [ makeWrapper emacs ];
```

File: pkgs/applications/editors/emacs-modes/emacspeak/default.nix [hunk 12]
```diff
@@ -0,0 +1,43 @@
+{ stdenv, fetchurl, fetchFromGitHub, makeWrapper, rsync, gnused, emacs, tcl, tclx, espeak-ng }:
```
Comment (Mic92) [inline pkgs/applications/editors/emacs-modes/emacspeak/default.nix:1]:
```suggestion
{ stdenv, fetchurl, fetchFromGitHub, makeWrapper, gnused, emacs, tcl, tclx, espeak-ng }:
```

**Review by Mic92 — CHANGES_REQUESTED** [review]

**Comment by Dema** [comment]
Forgot to push even more clean version :)

**Comment by Mic92** [comment]
This was working for me now:

```
nix-shell -p emacsPackages.emacspeak --run "emacspeak"
```

I head espeak reading out emacs.

**Comment by adisbladis** [comment]
The `meta` attribute in this derivation was wrong, fixed in https://github.com/NixOS/nixpkgs/commit/a2ad2bad4876cab4447ab6d9b29193ffd8b073cf.

Also @Dema it seems like you are not in the maintainers list. If you want to maintain packages in nixpkgs please add yourself to the maintainers list & re-add yourself to the derivation in a follow-up PR.

**Pull request merged**